### PR TITLE
Invalidate cache for SiteGround Speed Optimizer when coming soon options are changed

### LINKED ISSUE #48515: Add direct cache expiration on site visibility change
## Issue Description

For the initial launch coming soon mode, issue direct cache expiration / flush commands when the site visibility settings change for caches with known issues.

## Acceptance Criteria

Expire the following caches in [ComingSoonCacheInvalidator.php](https://github.com/woocommerce/woocommerce/blob/5dd7713346786c5874615b34a1d56f81a29b673f/plugins/woocommerce/src/Internal/ComingSoon/ComingSoonCacheInvalidator.php)

* [Siteground](https://wordpress.org/plugins/sg-cachepress/)
* [WPEngine](https://wpengine.com/support/cache/#Purge_Server_Caches) (see "Purge Cache with PHP")

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/48515.

Expire the following caches in [ComingSoonCacheInvalidator.php](https://github.com/woocommerce/woocommerce/blob/5dd7713346786c5874615b34a1d56f81a29b673f/plugins/woocommerce/src/Internal/ComingSoon/ComingSoonCacheInvalidator.php)

- [Siteground](https://wordpress.org/plugins/sg-cachepress/) (https://wordpress.org/support/topic/instructions-needed-helper-function-to-purge-everything/)

```php
/**
 * Public function to purge the entire cache.
 *
 * @since  5.7.14
 */
function sg_cachepress_purge_everything() {
	global $siteground_optimizer_loader;

	// Purge Dynamic cache if enabled.
	if ( Options::is_enabled( 'siteground_optimizer_enable_cache' ) ) {
		$siteground_optimizer_loader->supercacher->purge_cache();
	}

	// Purge Memcached if enabled.
	if ( Options::is_enabled( 'siteground_optimizer_enable_memcached' ) ) {
		$siteground_optimizer_loader->supercacher->flush_memcache();
	}

	if ( Options::is_enabled( 'siteground_optimizer_file_caching' ) ) {
		$siteground_optimizer_loader->file_cacher->purge_everything();
	}

	$siteground_optimizer_loader->supercacher->delete_assets();
}
```

- [WPEngine](https://wpengine.com/support/cache/#Purge_Server_Caches) (see "Purge Cache with PHP")

> Through the WP Engine MU plugin there is a function called wpecommon::purge_varnish_cache(). The post ID you want to purge can be passed into this function. Varnish page cache is purged only for that post URL, and not for the entire domain. This can have a positive impact on a website’s performance by keeping all other pages stored in cached.

> If wpecommon::purge_varnish_cache() is called without being passed a post ID, then Varnish will be purged for the entire domain.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Build the plugin zip file
2. Upload the plugin zip file to a SiteGround site (p1718618167417629/1718593364.831769-slack-C01SFMVEYAK)
3. Activate the plugin
4. Set the site to coming soon mode
5. Go to the site front end in a incognito window and confirm the site is in coming soon mode
6. Change the coming soon mode settings to Live
7. Confirm the cache is invalidated and the site is live

### Changelog entry

- [ ] Automatically create a changelog entry from the details below.

- [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [ ] Patch
- [ ] Minor
- [ ] Major

#### Type

- [ ] Fix - Fixes an existing bug
- [ ] Add - Adds functionality
- [ ] Update - Update existing functionality
- [ ] Dev - Development related task
- [ ] Tweak - A minor adjustment to the codebase
- [ ] Performance - Address performance issues
- [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by moon0326 — APPROVED** [review]
LGTM and tested well 👍 🚀 

I was wondering if it's a good idea to include other hosting providers' purge methods?

I've found https://gist.github.com/IlanVivanco/dd74c588aede2376fa94104414a2e253 very helpful. This script includes 12 purge methods.

Pre-approving ✅

File: plugins/woocommerce/src/Internal/ComingSoon/ComingSoonCacheInvalidator.php [hunk 1]
```diff
@@ -39,5 +39,15 @@ public function invalidate_caches() {
...
+
+		// Invalidate the SiteGround Speed Optimizer cache.
+		if ( function_exists( '\sg_cachepress_purge_cache' ) ) {
+			\sg_cachepress_purge_cache();
+		}
+
+		// Invalidate the WP Engine cache.
+		if ( class_exists( '\wpecommon' ) && method_exists( 'wpecommon', 'purge_varnish_cache' ) ) {
+			\wpecommon::purge_varnish_cache();
+		}
```
Comment (adrianduffell) [inline plugins/woocommerce/src/Internal/ComingSoon/ComingSoonCacheInvalidator.php]:
```suggestion
```

Suggest removing this as we couldn't confirm it works on an WPE account and don't know exactly what the behavior is.

**Review by adrianduffell — APPROVED** [review]
LGTM as a temporary solution! 🚀 

The SiteGround cache tested well for me but I couldn't get the WP Engine one to work consistently. I suggest we remove WPE.

File: plugins/woocommerce/changelog/update-purge-cache-siteground-wpengine [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: update
+
+Invalidate cache for SiteGround Speed Optimizer and WP Engine
```
Comment (chihsuan) [inline plugins/woocommerce/changelog/update-purge-cache-siteground-wpengine]:
```suggestion
Invalidate cache for SiteGround Speed Optimizer
```

**Pull request merged**